README.md
# Pheature Flags <br><sub><sup>Doctrine DBAL Toggle</sup></sub>

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Type coverage][ico-psalm]][link-psalm]
[![Test Coverage][ico-coverage]][link-coverage]
[![Mutation testing coverage][ico-mutant]][link-mutant]
[![Total Downloads][ico-downloads]][link-downloads]
[![Maintainability][ico-maintain]][link-maintain]

## Installation

Install it using [composer package manager](https://getcomposer.org/download/).

```php
composer require pheature/dbal-toggle
```

## Contributing

Your contributions are always welcome! Please have a look at the [contribution guidelines](./CONTRIBUTING.md) first.

## License

We really believe in the Open Source Software, we built our carers around it, and we feel that we need to return our
knowledge to the community. For this reason we release all our packages under [BSD-3-Clause licence](./LICENSE.md). 

[ico-code-quality]: https://img.shields.io/scrutinizer/g/pheature-flags/dbal-toggle.svg?style=flat-square
[link-code-quality]: https://scrutinizer-ci.com/g/pheature-flags/dbal-toggle/badges/coverage.png?b=1.0.x
[ico-version]: https://img.shields.io/packagist/v/pheature/dbal-toggle.svg?style=flat-square
[link-packagist]: https://packagist.org/packages/pheature/dbal-toggle
[ico-downloads]: https://img.shields.io/packagist/dt/pheature/dbal-toggle.svg?style=flat-square
[link-downloads]: https://packagist.org/packages/pheature/dbal-toggle
[ico-maintain]: https://api.codeclimate.com/v1/badges/a674ee0aad396e24bd7c/maintainability
[link-maintain]: https://codeclimate.com/github/pheature-flags/dbal-toggle/maintainability
[ico-psalm]: https://shepherd.dev/github/pheature-flags/dbal-toggle/coverage.svg
[link-psalm]: https://shepherd.dev/github/pheature-flags/dbal-toggle
[ico-mutant]: https://img.shields.io/endpoint?style=flat&url=https%3A%2F%2Fbadge-api.stryker-mutator.io%2Fgithub.com%2Fpheature-flags%2Fdbal-toggle%2F1.0.x
[link-mutant]: https://dashboard.stryker-mutator.io/reports/github.com/pheature-flags/dbal-toggle/1.0.x
[ico-coverage]: https://codecov.io/gh/pheature-flags/dbal-toggle/branch/1.0.x/graph/badge.svg?token=DTQIQUZ106
[link-coverage]: https://codecov.io/gh/pheature-flags/dbal-toggle


CHANGELOG.md
# Changelog

## [0.8.0](https://github.com/pheature-flags/dbal-toggle/tree/0.8.0)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/0.7.1...0.8.0)

**Implemented enhancements:**

- Improve assertion method call and fix Psalm issue [\#33](https://github.com/pheature-flags/dbal-toggle/pull/33) (@kpicaza)

**Merged pull requests:**

- update psalm to 5.16 [\#34](https://github.com/pheature-flags/dbal-toggle/pull/34) (@kpicaza)
- update php versions support [\#32](https://github.com/pheature-flags/dbal-toggle/pull/32) (@kpicaza)

## [0.7.1](https://github.com/pheature-flags/dbal-toggle/tree/0.7.1) (2022-12-12)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/0.7.0...0.7.1)

**Fixed bugs:**

- \[symfony-toggle\#44\] update dependency constraints to ^0.7 [\#30](https://github.com/pheature-flags/dbal-toggle/pull/30) (@kpicaza)

## [0.7.0](https://github.com/pheature-flags/dbal-toggle/tree/0.7.0) (2022-12-10)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/0.5.0...0.7.0)

**Merged pull requests:**

- \[\#359\] Add PHP 8.2 support [\#28](https://github.com/pheature-flags/dbal-toggle/pull/28) (@kpicaza)
- update dev dependencies [\#24](https://github.com/pheature-flags/dbal-toggle/pull/24) (@kpicaza)

## [0.5.0](https://github.com/pheature-flags/dbal-toggle/tree/0.5.0) (2022-08-12)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/0.4.0...0.5.0)

**Merged pull requests:**

- bump interdependencies to ^0.5 [\#23](https://github.com/pheature-flags/dbal-toggle/pull/23) (@kpicaza)

## [0.4.0](https://github.com/pheature-flags/dbal-toggle/tree/0.4.0) (2022-07-25)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/v0.3.1...0.4.0)

**Merged pull requests:**

- allow executing some composer plugins in dev mode [\#21](https://github.com/pheature-flags/dbal-toggle/pull/21) (@kpicaza)

## [v0.3.1](https://github.com/pheature-flags/dbal-toggle/tree/v0.3.1) (2022-03-01)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/0.3.0...v0.3.1)

**Implemented enhancements:**

- \[dbal-toggle\] Add CLI option to InitSchema to create schema if not exists [\#17](https://github.com/pheature-flags/dbal-toggle/issues/17)
- \[dbal-toggle\] Update issue template for bugs using new issue form in a YAML file [\#3](https://github.com/pheature-flags/dbal-toggle/issues/3)
- \[\#17\]\[dbal-toggle\] Add CLI option to InitSchema to create schema if not exists [\#18](https://github.com/pheature-flags/dbal-toggle/pull/18) (@xserrat)

**Fixed bugs:**

- \[dbal-toggle\] Throw DbalFeatureNotFound exception instead of InvalidArgumentException [\#19](https://github.com/pheature-flags/dbal-toggle/issues/19)
- \[\#19\]\[dbal-toggle\] Throw DbalFeatureNotFound exception instead of InvalidArgumentException [\#20](https://github.com/pheature-flags/dbal-toggle/pull/20) (@xserrat)

**Merged pull requests:**

- \[\#3\] Update Bug Report Template [\#14](https://github.com/pheature-flags/dbal-toggle/pull/14) (@pcs289)

## [0.3.0](https://github.com/pheature-flags/dbal-toggle/tree/0.3.0) (2021-11-21)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/0.2.0...0.3.0)

**Implemented enhancements:**

- \[dbal-toggle\] Update PhpStan to version 1 and level 9 [\#8](https://github.com/pheature-flags/dbal-toggle/issues/8)
- \[dbal-toggle\] Add PHP 8.1 support [\#6](https://github.com/pheature-flags/dbal-toggle/issues/6)
- \[\#6\] Add support for PHP 8.1 [\#12](https://github.com/pheature-flags/dbal-toggle/pull/12) (@kpicaza)
- \[\#8\] Update PhpStan to version 1 and level 9 [\#9](https://github.com/pheature-flags/dbal-toggle/pull/9) (@kpicaza)

**Closed issues:**

- \[dbal-toggle\] update CI matrix PHP versions [\#10](https://github.com/pheature-flags/dbal-toggle/issues/10)

**Merged pull requests:**

- \[\#10\] Run CI both PHP 7.4 and 8.0 [\#11](https://github.com/pheature-flags/dbal-toggle/pull/11) (@kpicaza)

## [0.2.0](https://github.com/pheature-flags/dbal-toggle/tree/0.2.0) (2021-10-08)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/v0.1.2...0.2.0)

**Implemented enhancements:**

- Update `DbalFeatureRepository::remove` contract with a `Feature` instead of `FeatureId` [\#4](https://github.com/pheature-flags/dbal-toggle/issues/4)
- \[\#4\] Update DbalFeatureRepository::remove contract with a Feature instead of FeatureId [\#5](https://github.com/pheature-flags/dbal-toggle/pull/5) (@xserrat)

## [v0.1.2](https://github.com/pheature-flags/dbal-toggle/tree/v0.1.2) (2021-09-25)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/v0.1.1...v0.1.2)

## [v0.1.1](https://github.com/pheature-flags/dbal-toggle/tree/v0.1.1) (2021-06-20)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/0.1.0...v0.1.1)

## [0.1.0](https://github.com/pheature-flags/dbal-toggle/tree/0.1.0) (2021-06-19)

[Full Changelog](https://github.com/pheature-flags/dbal-toggle/compare/2ffb3f5f713cafc343e7cb4030ccd9ed9644831f...0.1.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


test/DbalSchemaTest.php
<?php

declare(strict_types=1);

namespace Pheature\Test\Dbal\Toggle;

use Doctrine\DBAL\Connection;
use Doctrine\DBAL\DriverManager;
use Doctrine\DBAL\Schema\SchemaException;
use Pheature\Dbal\Toggle\DbalSchema;
use PHPUnit\Framework\TestCase;

class DbalSchemaTest extends TestCase
{
    private string $dbPath;
    private Connection $connection;

    protected function setUp(): void
    {
        $this->dbPath = realpath(__DIR__) . '/test.sqlite';
        touch($this->dbPath);
        $this->connection = DriverManager::getConnection(['url' => 'sqlite:///' . $this->dbPath]);
    }

    protected function tearDown(): void
    {
        unlink($this->dbPath);
    }

    public function testItShouldCreateValidDatabaseSchema(): void
    {
        $schema = new DbalSchema($this->connection);

        $initializeIfNotExists = false;
        $schema($initializeIfNotExists);

        $statement = $this->connection->executeQuery('SELECT `sql` FROM sqlite_master sm WHERE sm.name = "pheature_toggles"');

        $tableSchema = method_exists($statement, 'fetchOne') ? $statement->fetchOne() : $statement->fetch()['sql'];

        $this->assertStringContainsString('CREATE TABLE pheature_toggles', $tableSchema);
        $this->assertStringContainsString('feature_id VARCHAR(36) NOT NULL', $tableSchema);
        $this->assertStringContainsString('name VARCHAR(140) NOT NULL', $tableSchema);
        $this->assertStringContainsString('enabled BOOLEAN NOT NULL', $tableSchema);
        $this->assertStringContainsString('strategies CLOB DEFAULT \'[]\' NOT NULL --(DC2Type:json)', $tableSchema);
        $this->assertStringContainsString('created_at DATETIME NOT NULL --(DC2Type:datetime_immutable)', $tableSchema);
        $this->assertStringContainsString('updated_at DATETIME DEFAULT NULL --(DC2Type:datetime_immutable)', $tableSchema);
        $this->assertStringContainsString('PRIMARY KEY(feature_id))', $tableSchema);
    }

    public function testItShouldThrowAnExceptionIfSchemaTableAlreadyExists(): void
    {
        $schema = new DbalSchema($this->connection);

        $this->expectException(SchemaException::class);
        $this->expectExceptionMessage('The table with name "public.pheature_toggles" already exists.');

        $initializeIfNotExists = false;
        $schema($initializeIfNotExists);
        $schema($initializeIfNotExists);
    }

    public function testItShouldDoNothingIfSchemaTableAlreadyExistsSettingInitializeIfNotExistsToTrue(): void
    {
        $schema = new DbalSchema($this->connection);

        $initializeIfNotExists = true;
        $schema($initializeIfNotExists);
        $schema($initializeIfNotExists);

        $statement = $this->connection->executeQuery('SELECT `sql` FROM sqlite_master sm WHERE sm.name = "pheature_toggles"');

        $tableSchema = method_exists($statement, 'fetchOne') ? $statement->fetchOne() : $statement->fetch()['sql'];
        $this->assertStringContainsString('CREATE TABLE pheature_toggles', $tableSchema);
    }
}


test/Cli/InitSchemaTest.php
<?php

namespace Pheature\Test\Dbal\Toggle\Cli;

use Doctrine\DBAL\Connection;
use Doctrine\DBAL\Platforms\AbstractPlatform;
use Doctrine\DBAL\Schema\AbstractSchemaManager;
use Doctrine\DBAL\Schema\Schema;
use Doctrine\DBAL\Schema\Table;
use Pheature\Dbal\Toggle\Cli\InitSchema;
use Pheature\Dbal\Toggle\DbalSchema;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class InitSchemaTest extends TestCase
{
    public function testItShouldBeASymfonyCommand(): void
    {
        $schema = $this->createConfiguredMock(Schema::class, [
            'toSql' => [],
        ]);
        $schema->expects(self::once())
            ->method('createTable')
            ->with('pheature_toggles')
            ->willReturn($this->createMock(Table::class));
        $platform = $this->createMock(AbstractPlatform::class);

        if (method_exists(Connection::class, 'createSchemaManager')) {
            $schemaManager = $this->createConfiguredMock(AbstractSchemaManager::class, [
                'introspectSchema' => $schema,
            ]);
            $connection = $this->createConfiguredMock(Connection::class, [
                'createSchemaManager' => $schemaManager,
                'getDatabasePlatform' => $platform,
            ]);
        } else {
            $schemaManager = $this->createConfiguredMock(AbstractSchemaManager::class, [
                'createSchema' => $schema,
            ]);
            $connection = $this->createConfiguredMock(Connection::class, [
                'getSchemaManager' => $schemaManager,
                'getDatabasePlatform' => $platform,
            ]);
        }
        $initSchema = new InitSchema(new DbalSchema($connection));

        self::assertSame('pheature:dbal:init-toggle', $initSchema->getName());
        self::assertSame('Create Pheature toggles database schema.', $initSchema->getDescription());

        $input = $this->createMock(InputInterface::class);
        $output = $this->createMock(OutputInterface::class);
        $output->expects(self::once())
            ->method('writeLn')
            ->with('<info>Pheature Toggle database schema successfully created.</info>');
        self::assertSame(0, $initSchema->run($input, $output));
    }
}


test/DbalFeatureFactoryTest.php
<?php

declare(strict_types=1);

namespace Pheature\Test\Dbal\Toggle\Read;

use Pheature\Core\Toggle\Read\ChainToggleStrategyFactory;
use Pheature\Core\Toggle\Read\SegmentFactory;
use Pheature\Core\Toggle\Read\ToggleStrategyFactory;
use Pheature\Dbal\Toggle\Read\DbalFeatureFactory;
use Pheature\Model\Toggle\Feature;
use PHPUnit\Framework\TestCase;
use function json_encode;

final class DbalFeatureFactoryTest extends TestCase
{
    /** @dataProvider getFeatureData */
    public function testCreateReadModelFeatureInstance(array $featureData): void
    {
        $strategyFactory = new ChainToggleStrategyFactory(
            $this->createMock(SegmentFactory::class),
            $this->createConfiguredMock(ToggleStrategyFactory::class, [
                'types' => ['enable_by_matching_segment']
            ]),
        );
        $factory = new DbalFeatureFactory($strategyFactory);
        $feature = $factory->create($featureData);
        $this->assertInstanceOf(Feature::class, $feature);
    }

    public function getFeatureData(): array
    {
        return [
            'Feature without strategies' => [
                [
                    'feature_id' => 'some_feature',
                    'enabled' => true,
                    'strategies' => '[]'
                ],
            ],
            'Feature with a single enable by matching segment strategy' => [
                [
                    'feature_id' => 'some_feature',
                    'enabled' => 1,
                    'strategies' => json_encode([
                        [
                            'strategy_id' => 'enable_for_developers',
                            'strategy_type' => 'enable_by_matching_segment',
                            'segments' => [
                                [
                                    'segment_id' => 'is_developer',
                                    'segment_type' => 'strict_matching_segment',
                                    'criteria' => [
                                        'developer',
                                    ],
                                ],
                            ],
                        ],
                    ]),
                ],
            ],
        ];
    }
}


src/Cli/InitSchema.php
<?php

declare(strict_types=1);

namespace Pheature\Dbal\Toggle\Cli;

use Pheature\Dbal\Toggle\DbalSchema;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;

final class InitSchema extends Command
{
    private DbalSchema $dbalSchema;

    public function __construct(DbalSchema $dbalSchema)
    {
        parent::__construct();
        $this->dbalSchema = $dbalSchema;
    }

    protected function configure(): void
    {
        $this->setName('pheature:dbal:init-toggle')
            ->setDescription('Create Pheature toggles database schema.')
            ->addOption(
                'init-if-not-exists',
                null,
                InputOption::VALUE_NONE,
                'Initialize DB toggle schema if not exists'
            );
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $initializeIfNotExists = (bool)$input->getOption('init-if-not-exists');
        $this->dbalSchema->__invoke($initializeIfNotExists);

        $output->writeln('<info>Pheature Toggle database schema successfully created.</info>');

        return 0;
    }
}


src/Read/DbalFeatureFinder.php
<?php

declare(strict_types=1);

namespace Pheature\Dbal\Toggle\Read;

use Doctrine\DBAL\Connection;
use Doctrine\DBAL\Result;
use Pheature\Core\Toggle\Read\Feature;
use Pheature\Core\Toggle\Read\FeatureFinder;
use Pheature\Dbal\Toggle\Exception\DbalFeatureNotFound;

use function array_map;
use function is_array;

/**
 * @psalm-import-type DbalFeature from \Pheature\Dbal\Toggle\Read\DbalFeatureFactory
 */
final class DbalFeatureFinder implements FeatureFinder
{
    private Connection $connection;
    private DbalFeatureFactory $featureFactory;

    public function __construct(Connection $connection, DbalFeatureFactory $featureFactory)
    {
        $this->connection = $connection;
        $this->featureFactory = $featureFactory;
    }

    public function get(string $featureId): Feature
    {
        $sql = <<<SQL
        SELECT * FROM pheature_toggles WHERE feature_id = :feature_id
        SQL;

        $statement = $this->connection->executeQuery($sql, ['feature_id' => $featureId]);

        /** @var DbalFeature|false $feature */
        $feature = $statement->fetchAssociative();
        if (false === is_array($feature)) {
            throw DbalFeatureNotFound::withId($featureId);
        }

        return $this->featureFactory->create($feature);
    }

    /**
     * @return Feature[]
     * @throws \Doctrine\DBAL\Exception
     * @throws \JsonException
     */
    public function all(): array
    {
        $sql = <<<SQL
        SELECT * FROM pheature_toggles ORDER BY created_at DESC
        SQL;

        $statement = $this->connection->executeQuery($sql);

        return array_map(
            fn(array $feature) => $this->featureFactory->create($feature),
            $this->fetchFeatures($statement)
        );
    }

    /**
     * @return DbalFeature[]
     */
    private function fetchFeatures(Result $statement): array
    {
        /** @var DbalFeature[] $result */
        $result = $statement->fetchAllAssociative();

        return $result;
    }
}


src/Write/DbalFeatureFactory.php
<?php

declare(strict_types=1);

namespace Pheature\Dbal\Toggle\Write;

use Pheature\Core\Toggle\Write\Feature;
use Pheature\Core\Toggle\Write\FeatureId;
use Pheature\Core\Toggle\Write\Payload;
use Pheature\Core\Toggle\Write\Segment;
use Pheature\Core\Toggle\Write\SegmentId;
use Pheature\Core\Toggle\Write\SegmentType;
use Pheature\Core\Toggle\Write\Strategy;
use Pheature\Core\Toggle\Write\StrategyId;
use Pheature\Core\Toggle\Write\StrategyType;

use function array_map;
use function json_decode;

/**
 * @psalm-import-type WriteStrategy from \Pheature\Core\Toggle\Write\Strategy
 * @psalm-import-type DbalFeature from \Pheature\Dbal\Toggle\Read\DbalFeatureFactory
 */
final class DbalFeatureFactory
{
    /**
     * @param DbalFeature $featureData
     * @return Feature
     */
    public static function createFromDbalRepresentation(array $featureData): Feature
    {
        /** @var WriteStrategy[] $strategiesData */
        $strategiesData = json_decode($featureData['strategies'], true, 12, JSON_THROW_ON_ERROR);
        /** @var callable $strategyCallback */
        $strategyCallback = static function (array $strategy): Strategy {
            /** @var WriteStrategy $strategy */
            $segments = array_map(
                static function (array $segment): Segment {
                    return new Segment(
                        SegmentId::fromString($segment['segment_id']),
                        SegmentType::fromString($segment['segment_type']),
                        Payload::fromArray($segment['criteria'])
                    );
                },
                $strategy['segments']
            );

            return new Strategy(
                StrategyId::fromString($strategy['strategy_id']),
                StrategyType::fromString($strategy['strategy_type']),
                $segments
            );
        };

        /** @var Strategy[] $strategies */
        $strategies = array_map($strategyCallback, $strategiesData);

        return new Feature(
            FeatureId::fromString($featureData['feature_id']),
            (bool)$featureData['enabled'],
            $strategies
        );
    }
}


src/Write/DbalFeatureRepository.php
<?php

declare(strict_types=1);

namespace Pheature\Dbal\Toggle\Write;

use Doctrine\DBAL\Connection;
use InvalidArgumentException;
use Pheature\Core\Toggle\Write\Feature;
use Pheature\Core\Toggle\Write\FeatureId;
use Pheature\Core\Toggle\Write\FeatureRepository;
use DateTimeImmutable;

use function json_encode;

/**
 * @psalm-import-type WriteStrategy from \Pheature\Core\Toggle\Write\Strategy
 * @psalm-import-type DbalFeature from \Pheature\Dbal\Toggle\Read\DbalFeatureFactory
 */
final class DbalFeatureRepository implements FeatureRepository
{
    private const TABLE = 'pheature_toggles';
    private Connection $connection;

    public function __construct(Connection $connection)
    {
        $this->connection = $connection;
    }

    public function save(Feature $feature): void
    {
        $now = new DateTimeImmutable();

        if (null === $this->findFeature($feature->id())) {
            $this->connection->insert(
                self::TABLE,
                [
                    'feature_id' => $feature->id(),
                    'name' => $feature->id(),
                    'enabled' => (int)$feature->isEnabled(),
                    'strategies' => json_encode($feature->strategies(), JSON_THROW_ON_ERROR),
                    'created_at' => $now->format('Y-m-d H:i:s'),
                ]
            );
            return;
        }

        $this->connection->update(
            self::TABLE,
            [
                'enabled' => (int)$feature->isEnabled(),
                'strategies' => json_encode($feature->strategies(), JSON_THROW_ON_ERROR),
                'updated_at' => $now->format('Y-m-d H:i:s'),
            ],
            [
                'feature_id' => $feature->id(),
            ]
        );
    }

    public function get(FeatureId $featureId): Feature
    {
        $featureData = $this->findFeature($featureId->value());
        if (null === $featureData) {
            throw new InvalidArgumentException(sprintf('Not feature found for given id %s', $featureId->value()));
        }

        return DbalFeatureFactory::createFromDbalRepresentation($featureData);
    }

    public function remove(Feature $feature): void
    {
        $this->connection->delete(
            self::TABLE,
            [
                'feature_id' => $feature->id()
            ]
        );
    }

    /**
     * @return DbalFeature|null
     * @throws \Doctrine\DBAL\Exception
     */
    private function findFeature(string $id): ?array
    {
        $table = self::TABLE;
        $sql = <<<SQL
            SELECT * FROM $table WHERE feature_id = :feature_id
        SQL;

        $statement = $this->connection->executeQuery($sql, ['feature_id' => $id]);
        /** @var DbalFeature|false $result */
        $result = $statement->fetchAssociative();

        return $result ?: null;
    }
}


src/DbalSchema.php
<?php

declare(strict_types=1);

namespace Pheature\Dbal\Toggle;

use Doctrine\DBAL\Connection;
use Doctrine\DBAL\Platforms\AbstractPlatform;
use Doctrine\DBAL\Schema\Schema;
use Doctrine\DBAL\Schema\SchemaConfig;

final class DbalSchema
{
    private const TABLE_NAME = 'pheature_toggles';
    private Schema $schema;
    private AbstractPlatform $platform;
    private Connection $connection;

    public function __construct(Connection $connection)
    {
        if (method_exists($connection, 'createSchemaManager')) {
            $this->schema = $connection->createSchemaManager()->introspectSchema();
        } else {
            /** @psalm-suppress DeprecatedMethod */
            $this->schema = $connection->getSchemaManager()->createSchema();
        }
        $this->platform = $connection->getDatabasePlatform();
        $this->connection = $connection;
    }

    public function __invoke(bool $initializeIfNotExists): void
    {
        if (!$this->shouldBeExecuted($initializeIfNotExists)) {
            return;
        }

        $this->createPheatureTogglesTable();

        $queries = $this->schema->toSql($this->platform);
        foreach ($queries as $query) {
            if (false !== strpos($query, self::TABLE_NAME)) {
                $this->connection->executeQuery($query);
            }
        }
    }

    private function createPheatureTogglesTable(): void
    {
        $table = $this->schema->createTable(self::TABLE_NAME);
        $table->addColumn(
            'feature_id',
            'string',
            [
                'length' => 36,
            ]
        );
        $table->setPrimaryKey(['feature_id']);
        $table->addColumn(
            'name',
            'string',
            [
                'length' => 140,
            ]
        );
        $table->addColumn('enabled', 'boolean');
        if ($this->isSqlite()) {
            $table->addColumn(
                'strategies',
                'json',
                [
                    'default' => '[]'
                ]
            );
        } else {
            $table->addColumn('strategies', 'json');
        }
        $table->addColumn('created_at', 'datetime_immutable');
        $table->addIndex(['created_at']);
        $table->addColumn(
            'updated_at',
            'datetime_immutable',
            [
                'notnull' => false,
                'default' => null,
            ]
        );
    }

    private function shouldBeExecuted(bool $initializeIfNotExists): bool
    {
        if (!$initializeIfNotExists) {
            return true;
        }

        if ($this->schema->hasTable(self::TABLE_NAME)) {
            return false;
        }

        return true;
    }

    /**
     * @psalm-suppress MixedInferredReturnType
     */
    private function isSqlite(): bool
    {
        $driver = $this->connection->getDriver();

        /**
         * @psalm-suppress UndefinedClass
         */
        if (
            class_exists(\Doctrine\DBAL\Driver\PDOSqlite\Driver::class)
            && $driver instanceof \Doctrine\DBAL\Driver\PDOSqlite\Driver
        ) {
            return true;
        }

        if (
            class_exists(\Doctrine\DBAL\Driver\PDO\SQLite\Driver::class)
            && $driver instanceof \Doctrine\DBAL\Driver\PDO\SQLite\Driver
        ) {
            return true;
        }

        if (
            class_exists(\Doctrine\DBAL\Driver::class)
            && $driver instanceof \Doctrine\DBAL\Driver
        ) {
            /**
             * @psalm-suppress DeprecatedMethod
             */
            return 'sqlite' === $driver->getDatabasePlatform()->getName();
        }

        return false;
    }
}
